dialer.go

```diff
@@ -213,12 +218,22 @@ func (d *Dialer) dialContext(ctx context.Context, network string, address string
 			address = net.JoinHostPort(address, port)
 		}
 	}
-	return (&net.Dialer{
+
+	conn, err := (&net.Dialer{
 		LocalAddr:     d.LocalAddr,
 		DualStack:     d.DualStack,
 		FallbackDelay: d.FallbackDelay,
 		KeepAlive:     d.KeepAlive,
 	}).DialContext(ctx, network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	if d.TLS != nil {
+		return tls.Client(conn, d.TLS), nil
```

Comment:
Doing this delays the TLS handshake until a message is first read or written, which may have unexpected consequences (like timeouts if the server expects the handshake to come quickly, but the first message may be produced too late).

This is what I suggested to do the handshake in a separate goroutine, and select on the context and the completion of that goroutine to support proper cancellation during this step.

Reply:
Ah, gotcha.  Didn't realize this delayed the handshake.  Let me take another stab at it.  

On a separate note, I took at crack at consumer groups yesterday.  It's far from finished, but it's probably far enough along to get your feedback on it.  https://github.com/savaki/kafka-go/tree/consumer-groups

Currently, it handles in a very basic form: find coordinator, join group, and sync group